code/socialDistribution/views.py

```diff
@@ -510,21 +510,59 @@ def share_post(request, id):
     """
     if request.method == 'POST':
         author = LocalAuthor.objects.get(user=request.user)
-        post = LocalPost.objects.get(id=id)
+        
+        try:
+            post = get_object_or_404(LocalPost, id=id)
+            if not post.is_public() and not post.is_friends():
+                return redirect('socialDistribution:home')
+            # origin remains unchanged as the original true 'source'
+            oldSource = post.get_id()
+
+            post.pk = None  # duplicate the post
+            post.author = author
+            post.published = timezone.now()
+            post.source = oldSource
+            post.save()
+            to_dispatch = post
+            
+        except:
+            oldPost = get_object_or_404(InboxPost, id=id)
+            if not oldPost.is_public() and not oldPost.is_friends():
+                return redirect('socialDistribution:home')
+            new_post = LocalPost(
+                    author_id=author.id,
+                    title=oldPost.title,
+                    description=oldPost.description,
+                    content_type=oldPost.content_type,
+                    content=oldPost.content,
+                    visibility=oldPost.visibility,
+                    unlisted=oldPost.unlisted)
+
+            # set post origin and source to itself for a new post
+            new_post.origin = new_post.source = new_post.get_id()
```

Comment:
Shouldn't `origin` be the origin of the original inbox post, and `source` be the public_id of the original inbox post

Reply:
Fixed thanks